src/mlpack/core/tree/cover_tree/cover_tree_impl.hpp

```diff
@@ -547,6 +548,65 @@ CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::CoverTree(
   }
 }
 
+// Copy Assignment.
+template<
+    typename MetricType,
+    typename StatisticType,
+    typename MatType,
+    typename RootPointPolicy
+>
+CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>&
+CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
+operator=(const CoverTree& other)
+{
+  if (this == &other)
+    return *this;
+
+  dataset = ((other.parent == NULL && other.localDataset) ?
```

Comment:
Sorry, I just noticed something in the other PR that we should fix here before merging.  In the copy and move constructor we must first make sure to free any memory that will not be used anymore.  Here that will include `dataset` and `children`.

Reply:
Did you mean copy/move assignment or constructor? Or both?

And sorry if this seems like a silly question but would I go about doing this by inserting something like `if(dataset)  delete dataset;` ?

Reply:
Right, exactly, in the constructors it is not possible that memory has already been allocated for members like `dataset`.

Actually the `if (dataset)` check is not even necessary, because `delete` already performs the check:

https://stackoverflow.com/questions/4190703/is-it-safe-to-delete-a-null-pointer

So you can just do, e.g., `delete dataset;`.  However, note that this does *not* set `dataset` to `NULL`, so that is something to be aware of.